### PR TITLE
Handle import refs in TryGetExistingDecl

### PR DESCRIPTION
**Pull request by ozguronsoy**
This solves the problem where `interface` imports are incorrectly diagnosed as duplicate names in impl files.
Follows the implementation logic used in `handle_class.cpp`.

### CODE REVIEW

**Comment by lnihlen** [comment]
Overall LGTM but I'd like someone more familiar with check to give the final approval, so I'm tapping @zygoloid in for help on this one.

File: toolchain/check/testdata/interface/no_definition_in_impl_file.carbon [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+library "[[@TEST_NAME]]";
+
+interface A;
+
+// --- decl_in_api_definition_in_impl.impl.carbon
+
+impl library "[[@TEST_NAME]]";
+
+interface A;
```
Comment (zygoloid) [inline toolchain/check/testdata/interface/no_definition_in_impl_file.carbon:26]: I *think* this should be invalid under the rules in [#3762](https://docs.carbon-lang.dev/proposals/p003762-merging-forward-declarations.html): this declaration is redundant due to the declaration in the API file, and so is not permitted. The rules don't seem very explicit on this, but they do say:

> A declaration should always add new information.

Maybe rename to `todo_fail_`?

File: toolchain/check/testdata/interface/fail_definition_imported.carbon [hunk 2]
```diff
@@ -23,11 +23,11 @@ interface I;
 library "[[@TEST_NAME]]";
 import library "a";
 
-// CHECK:STDERR: fail_b.carbon:[[@LINE+8]]:11: error: duplicate name `I` being declared in the same scope [NameDeclDuplicate]
+// CHECK:STDERR: fail_b.carbon:[[@LINE+8]]:1: error: redeclaration of `interface I` is redundant [RedeclRedundant]
```
Comment (zygoloid) [inline toolchain/check/testdata/interface/fail_definition_imported.carbon:27]: I think this error is less good than the old one, but neither is really describing the problem. Ideally we'd say something more like: "error: name `I` declared in multiple libraries".

Maybe add a TODO comment?

File: toolchain/check/interface.cpp [hunk 3]
```diff
@@ -204,39 +212,94 @@ auto TryGetExistingDecl(Context& context, const NameComponent& name,
...
+    }
+    case CARBON_KIND(SemIR::ImportRefLoaded import_ref): {
+      auto import_ir_inst =
+          context.import_ir_insts().Get(import_ref.import_ir_inst_id);
+
+      // Verify the decl so that things like aliases are name conflicts.
+      const auto* import_ir =
+          context.import_irs().Get(import_ir_inst.ir_id()).sem_ir;
+      if (!import_ir->insts()
+               .IsOneOf<SemIR::InterfaceDecl, SemIR::InterfaceWithSelfDecl>(
```
Comment (zygoloid) [inline toolchain/check/interface.cpp]: We should check for `InterfaceDecl` here for an interface, and `NamedConstraintDecl` for a named constraint.

File: toolchain/check/interface.cpp [hunk 4]
```diff
@@ -204,39 +212,94 @@ auto TryGetExistingDecl(Context& context, const NameComponent& name,
     return std::nullopt;
   }
 
-  SemIR::InstId existing_id = lookup_result.target_inst_id();
-  SemIR::Inst existing_decl_inst = context.insts().Get(existing_id);
-  const auto* existing_decl_entity =
-      TryGetEntity<EntityT>(context, existing_decl_inst);
-  if (!existing_decl_entity) {
+  auto prev_id = lookup_result.target_inst_id();
```
Comment (zygoloid) [inline toolchain/check/interface.cpp:193]: This looks very similar to the logic in `handle_class.cpp`. Is it feasible to factor this out and share the code? (Maybe not, but it seems worth checking.)

File: toolchain/check/interface.cpp [hunk 5]
```diff
@@ -204,39 +212,94 @@ auto TryGetExistingDecl(Context& context, const NameComponent& name,
...
+
+  if (!prev_entity_id.has_value()) {
     // This is a redeclaration with a different entity kind.
     DiagnoseDuplicateName(context, name.name_id, name.name_loc_id,
-                          SemIR::LocId(existing_id));
+                          SemIR::LocId(prev_id));
     return std::nullopt;
   }
 
+  auto&& prev_entity = [&]() -> decltype(auto) {
```
Comment (zygoloid) [inline toolchain/check/interface.cpp]:
```suggestion
  const auto& prev_entity = [&]() -> const EntityT& {
```
We're never actually moving here, so using `auto&&` is misleading. I'm also inclined to avoid returning `decltype(auto)` given we can fairly easily spell the intended return type, just for clarity.

File: toolchain/check/interface.cpp
Comment (zygoloid) [inline toolchain/check/interface.cpp]: Can these be deleted? I think  you removed the only call to them.

File: toolchain/check/testdata/interface/no_definition_in_impl_file.carbon (hunk 1 above)
Comment (ozguronsoy) [reply toolchain/check/testdata/interface/no_definition_in_impl_file.carbon:26]: We also have the same test in `class/no_definition_in_impl_file.carbon` and `function/declaration/no_definition_in_impl_file.carbon`. Should I rename them all?

File: toolchain/check/testdata/interface/fail_definition_imported.carbon (hunk 2 above)
Comment (ozguronsoy) [reply toolchain/check/testdata/interface/fail_definition_imported.carbon:27]: Done.

File: toolchain/check/interface.cpp (hunk 3 above)
Comment (ozguronsoy) [reply toolchain/check/interface.cpp]: Oops, I initially implemented this for interface only and forgot to update it later.

Done.

File: toolchain/check/interface.cpp
Comment (ozguronsoy) [reply toolchain/check/interface.cpp]: Done.

File: toolchain/check/interface.cpp (hunk 4 above)
Comment (ozguronsoy) [reply toolchain/check/interface.cpp:193]: We also have a similar logic in `handle_function.cpp`. Maybe we can include `SemIR::Class` and `SemIR::Function` in `TryGetExistingDecl` and move it to some other file?

File: toolchain/check/testdata/interface/no_definition_in_impl_file.carbon (hunk 1 above)
Comment (zygoloid) [reply toolchain/check/testdata/interface/no_definition_in_impl_file.carbon:26]: Yes, please. Thanks for checking!

File: toolchain/check/interface.cpp (hunk 4 above)
Comment (zygoloid) [reply toolchain/check/interface.cpp:193]: Yeah, that'd work for me. If you prefer, I'm happy for this to land with the code duplicated; we can do the refactoring separately.

File: toolchain/check/testdata/interface/no_definition_in_impl_file.carbon (hunk 1 above)
Comment (ozguronsoy) [reply toolchain/check/testdata/interface/no_definition_in_impl_file.carbon:26]: Done.

File: toolchain/check/interface.cpp (hunk 4 above)
Comment (ozguronsoy) [reply toolchain/check/interface.cpp:193]: Thanks! I'll do the refactoring in a follow-up PR.

File: toolchain/check/testdata/interface/no_definition_in_impl_file.carbon [hunk 6]
```diff
@@ -0,0 +1,224 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/none.carbon
+// TODO: Add ranges and switch to "--dump-sem-ir-ranges=only".
+// EXTRA-ARGS: --dump-sem-ir-ranges=if-present
```
Comment (zygoloid) [inline toolchain/check/testdata/interface/no_definition_in_impl_file.carbon:7]: We are trying to remove these flags over time; it'd be preferable to not add new ones.

* If you think the `CHECK:STDOUT:` lines below are valuable, then add `//@dump-sem-ir-begin` and `//@dump-sem-ir-end` directives around the parts of the test below that you would like to be dumped.
* If you think the `CHECK:STDOUT:` lines aren't useful, then just delete this `EXTRA-ARGS:` line and the TODO comment, and they'll go away on autoupdate.

For this test I don't think the SemIR dump is providing any real value so I'd lean towards removing it.

**Review by zygoloid — APPROVED** [review]
Looks good! Thank you for working on this!

One trivial thing with the tests, but given that the same comment applies to the existing tests for classes and functions too, it's probably best left to the refactoring PR, so I'm going to merge this as-is.

File: toolchain/check/testdata/interface/no_definition_in_impl_file.carbon (hunk 6 above)
Comment (ozguronsoy) [reply toolchain/check/testdata/interface/no_definition_in_impl_file.carbon:7]: Thanks, I'll keep this in mind henceforth.
